utils/treesitter/src/scanner.c

```diff
@@ -54,34 +82,98 @@ static bool is_whitespace(char c) { return c == ' ' || c == '\n'; }
+    __attribute__((unused)) TSLexer* lexer, const bool* valid_symbols) {
   // skip past whitespace if any
   bool whitespace = false;
   while (is_whitespace(lexer->lookahead)) {
     whitespace = true;
     lexer->advance(lexer, /* skip= */ true);
   }
 
-  // if any other symbol than *, fallback to treesitter internal lexer
-  if (lexer->lookahead != '*') {
-    return false;
-  }
+  if (lexer->lookahead == '*' &&
+      (valid_symbols[BINARY_STAR] || valid_symbols[POSTFIX_STAR])) {
+    // move to past the *, add * to current token
+    lexer->advance(lexer, /* skip= */ false);
 
-  // move to past the *, add * to current token
-  lexer->advance(lexer, /* skip= */ false);
-
-  // https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/lexical_conventions/symbolic_tokens.md
-  if (is_whitespace(lexer->lookahead) && whitespace) {
-    // foo * bar
-    lexer->result_symbol = BINARY_STAR;
-  } else if (!whitespace &&
-             token_allowed_after_binary_operator(lexer->lookahead)) {
-    // foo*bar or foo*(bar)
-    lexer->result_symbol = BINARY_STAR;
-  } else {
-    // foo*
-    lexer->result_symbol = POSTFIX_STAR;
+    // https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/lexical_conventions/symbolic_tokens.md
+    if (is_whitespace(lexer->lookahead) && whitespace) {
+      // foo * bar
+      lexer->result_symbol = BINARY_STAR;
+    } else if (!whitespace &&
+               token_allowed_after_binary_operator(lexer->lookahead)) {
+      // foo*bar or foo*(bar)
+      lexer->result_symbol = BINARY_STAR;
+    } else {
+      // foo*
+      lexer->result_symbol = POSTFIX_STAR;
+    }
+    return true;
+  } else if ((lexer->lookahead == '#' || lexer->lookahead == '\'' ||
+              lexer->lookahead == '\"') &&
+             valid_symbols[STRING]) {
+    lexer->result_symbol = STRING;
+    int hash_count = 0;
+    while (lexer->lookahead == '#') {
+      advance(lexer);
+      hash_count++;
+    }
+    if (lexer->lookahead == '\'') {
+      if (!eat(lexer, "'''")) {
+        return false;
+      }
+      while (!lexer->eof(lexer)) {
+        // ''' must be on its own line.
+        // so look for \n *'''#{hash_count}
+        if (lexer->lookahead == '\n') {
+          advance(lexer);
+          while (lexer->lookahead == ' ') {
+            advance(lexer);
+          }
+          if (!eat(lexer, "'''")) {
+            continue;
+          }
+          if (eat_count(lexer, '#', hash_count)) {
+            // end of string
+            return true;
+          } else {
+            continue;
+          }
+        } else {
+          // skip not new lines
+          advance(lexer);
+        }
+      }
+    } else if (lexer->lookahead == '"') {
+      advance(lexer);
+      while (!lexer->eof(lexer)) {
+        if (lexer->lookahead == '\\') {
+          advance(lexer);
+          if (eat_count(lexer, '#', hash_count)) {
+            // skip the next character
```

Comment:
I'd much rather the comment say why than say what the code does. It isn't surprising that `advance(lexer)` skips a character, it is more surprising that it is needed.